sol/main.c

```diff
@@ -0,0 +1,513 @@
+		req_queue->lowest_priority = priority;
+	}
+
+	for (next = priority + 1; next < req_queue->num_priorities; next++) {
+		if (rte_bitmap_get(req_queue->bmp, next) != 0) {
+			pr->next = req_queue->priorities[next];
+			pr->prev = pr->next->prev;
+			pr->next->prev = pr;
+			if (pr->prev != NULL)
+				pr->prev->next = pr;
+			return;
+		}
+	}
+}
+
+static inline void
+insert_req(struct priority_req *last_req_of_pri, struct priority_req *new_req)
+{
+	struct priority_req *prev = last_req_of_pri->prev;
+	last_req_of_pri->prev = new_req;
+	new_req->next = last_req_of_pri;
+	if (prev) {
+		prev->next = new_req;
+		new_req->prev = prev;
+	}
+}
+
+static void
+drop_lowest_priority_pkt(struct sol_config *sol_conf,
+	struct req_queue *req_queue)
+{
+	struct priority_req *lowest_pr =
+		req_queue->priorities[req_queue->lowest_priority];
+
+	/* Only one packet in the queue. */
+	if (unlikely(lowest_pr->next == NULL)) {
+		req_queue->priorities[lowest_pr->priority] = NULL;
+		rte_bitmap_clear(req_queue->bmp, lowest_pr->priority);
+		req_queue->lowest_priority = req_queue->num_priorities;
+		goto out;
+	}
+
+	/* The lowest priority packet was the only one of that priority. */
+	if (lowest_pr->priority != lowest_pr->next->priority) {
+		req_queue->priorities[lowest_pr->priority] = NULL;
+		rte_bitmap_clear(req_queue->bmp, lowest_pr->priority);
+		lowest_pr->next->prev = NULL;
+		req_queue->lowest_priority = lowest_pr->next->priority;
+		goto out;
+	}
+
+	req_queue->priorities[lowest_pr->priority] = lowest_pr->next;
+	lowest_pr->next->prev = NULL;
+
+out:
+	rte_pktmbuf_free(lowest_pr->pkt);
+	mb_free_entry(&sol_conf->mb, lowest_pr);
+	req_queue->len--;
+}
+
+static void
+req_enqueue(struct sol_config *sol_conf, struct priority_req *req)
+{
+	struct req_queue *req_queue = sol_conf->req_queue;
+	uint8_t priority = req->priority;
+
+	if (req_queue->len == sol_conf->pri_req_max_len) {
+		/* New packet is lowest priority, so drop it. */
+		if (req_queue->lowest_priority == priority) {
+			rte_pktmbuf_free(req->pkt);
+			mb_free_entry(&sol_conf->mb, req);
+			return;
+		}
+		drop_lowest_priority_pkt(sol_conf, req_queue);
+	}
+
+	if (req_queue->priorities[priority] == NULL) {
+		/* Insert request of a priority we don't yet have. */
+		insert_new_priority_req(req_queue, req, priority);
+		rte_bitmap_set(req_queue->bmp, priority);
+	} else {
+		/* Append request to end of appropriate priority. */
+		insert_req(req_queue->priorities[priority], req);
+		req_queue->priorities[priority] = req;
+	}
+
+	req_queue->len++;
+}
+
+#define GK_REQ_ENQ_BURST_SIZE (32)
+#define GK_REQ_DEQ_BURST_SIZE (32)
+
+static void
+enqueue_reqs(struct sol_config *sol_conf)
+{
+	struct priority_req *req_nodes[GK_REQ_ENQ_BURST_SIZE];
+	int num_reqs = mb_dequeue_burst(&sol_conf->mb,
+		(void **)req_nodes, GK_REQ_ENQ_BURST_SIZE);
+	int i;
+	for (i = 0; i < num_reqs; i++)
+		req_enqueue(sol_conf, req_nodes[i]);
+}
+
+/* Scaling for cycles_per_byte calculation
```

Comment:
The first line of the comment should be ``/*``. It needs a period at the end of the comment.